geo_autoRIFT/autoRIFT/autoRIFT.py

```diff
@@ -28,229 +28,239 @@
 #
 # Note: this is based on the MATLAB code, "auto-RIFT", written by Alex S. Gardner,
 #       and has been translated to Python and further optimized.
-#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+import random
 
-
-
-import pdb
-import subprocess
-import re
-import string
+import cv2
 import sys
-
+import numpy as np
+from scipy.ndimage.morphology import distance_transform_edt
+from scipy.ndimage.filters import generic_filter
 
 
 class autoRIFT:
-    '''
+    """
     Class for mapping regular geographic grid on radar imagery.
-    '''
-    
-    
-    
+    """
+
+    def preprocess_filt_std(self, image):
+        filter_kernel = np.ones((self.WallisFilterWidth, self.WallisFilterWidth))
+        n = np.sum(filter_kernel)
+
+        shifted = image - np.mean(image)
+        conv_sum = cv2.filter2D(shifted, -1, filter_kernel, borderType=cv2.BORDER_REFLECT)
+        conv_squared_sum = cv2.filter2D(shifted ** 2, -1, filter_kernel, borderType=cv2.BORDER_REFLECT)
+
+        variance = (conv_squared_sum - ((conv_sum ** 2) / n)) / n
+        std = np.sqrt(variance)
+        return std
+
+    def preprocess_filt_wal_nodata_fill(self):
+        """
+        Wallis filter with nodata infill for L7 SLC Off preprocessing
+        """
+        self.zeroMask = np.full(self.I1.shape, True)
+        for image in (self.I1, self.I2):
+            zero_mask_1 = image == 0
+            buff = np.sqrt(2 * ((self.WallisFilterWidth - 1) / 2) ** 2) + 0.01
+
+            # find edges of image, this makes missing scan lines valid and will
+            # later be filled with random white noise
+            missing_data = distance_transform_edt(zero_mask_1) < 30
+            missing_data = missing_data & zero_mask_1
+            missing_data = distance_transform_edt(~missing_data) <= buff
+
+            # trying to frame out the image
+            valid_domain = ~zero_mask_1 | missing_data
+            self.zeroMask &= ~valid_domain
+
+            std = self.preprocess_filt_std(image)
+            print('using new version...')
+            low_std = std < self.StandardDeviationCutoff
+            low_std = distance_transform_edt(~low_std) <= buff
+            missing_data = (missing_data | low_std) & valid_domain
+
+            # apply wallis filter
+            kernel = np.ones((self.WallisFilterWidth, self.WallisFilterWidth), dtype=np.float32)
+            mean = cv2.filter2D(image, -1, kernel, borderType=cv2.BORDER_CONSTANT) / np.sum(kernel)
+            image = (image - mean) / std
+
+            valid_data = valid_domain & ~missing_data
+            image[~valid_data] = 0
```

Comment:
@alex-s-gardner at this point in the code, `image` pixel values should be mapped to a -1 to 1 range, correct? Since further down we're filling with Gaussian noise. 

Looking at the image here, the values range from -5.2 to 277.8, with about 22% of the valid data outside of this range.

Reply:
Nevermind, I can maths, I swear.